enhancements/mco-kargs-day1-support.md

```diff
@@ -0,0 +1,347 @@
+  the update without incurring an extra reboot.
+
+### Non-Goals
+
+- Does not support defining custom kernel arguments on non-RHCOS machines.
+- Does not support applying kernel args on first boot for nodes in clusters
+  installed using a 4.1 or earlier installer.
+- Does not protect against setting kernel arguments that may affect the
+  performance of the node or cluster.
+- Does not take into account kernel args set manually on the node when applying
+  kernel args from the MachineConfigs. Any kernel args manually set or unset
+  will be overwritten on the next update.
+
+## Proposal
+
+#### Part 1
+Add kernel args to install-config and process them in the MCO bootstrap. This
+should be a straight-forward translation from the install-config into
+MachineConfig resources.
+
+#### Part 2
+For the nuts and bolts, there are a couple ways we could achieve no-extra-reboot
+kernel args support.
+
+##### Method 1 - append to `etc/pivot/kernel-args`
+
+- Update the MachineConfigServer to parse kernel args in the MachineConfigs and
+  append them in the `/etc/pivot/kernel-args` file within the ignition it
+  serves, similar to how hyperthreading can be disabled in 4.2.
+- Update the MachineConfigDaemon to compare k-args set on MachineConfigs with
+  those set via rpm-ostree to determine if there are changes requiring a reboot.
+
+Pros:
+- Requires no additional packaging in RHCOS or other tricky methods to update
+  what's available on the host.
+- Allows new nodes in clusters with 4.2 bootimages to have k-args set without
+  the additional reboot.
+- Brings the MCD closer to the kubernetes model of inspecting current state and
+  reconciling desired spec with that.
+
+Cons:
+- Further special-cases kernel args as state passed to pivot. Any other config
+  that should be set before os pivot would need to be specially handled as well.
+- Determining the current state from the host rather than the current config
+  could get tricky, and is different from how other config fields are handled.
+  See https://github.com/openshift/machine-config-operator/pull/245
+
+
+##### Method 2 - MCD-as-pivot packaged in RHCOS
+
+- Installer in 4.2 ships machine-config-daemon package in RHCOS host which handles [early pivot](https://github.com/openshift/machine-config-operator/pull/859).
+- [MCS inverts MachineConfig](https://github.com/openshift/machine-config-operator/pull/868) containing non-ignition part (like osImageURL,kernelArguments) during first boot in /etc/ignition-machine-config-encapsulated.json on hosts
+- During the first boot, MCD is patched to parse the ignition-machine-config-encapsulated.json available on the host to read and set the k-args if specified.
+- machine-config-dameon package in RHCOS is updated so that it has required changes on the host during early boot.
+- MCS is updated so that MCD first boot service runs to process kargs
+- As a result, we get kargs applied on RHCOS nodes during first boot without incurring extra reboot.
```

Comment:
I think we've been long using `incurring in an extra reboot` wrongly. What we are really doing here is avoiding getting kubernetes up and running to just bring it down again. The "early pivot" phase effectively prevents that. Maybe this is easier to understand cc @cgwalters wdyt?

Reply:
I think it's both - we want to avoid double reboot, but we *also* want to avoid joining the cluster (and potentially landing workloads) only to immediately start to drain.